src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -458,6 +460,8 @@ data EvalEnv = EvalEnv
                                     -- an expression
   , evSMTCache :: M.HashMap Expr Bool -- ^ Whether an expression is valid or its negation
   , evFuel     :: FuelCount
+  , unfoldedDefinitions :: S.HashSet Symbol -- ^ Functions that we have already unfolded
```

Comment:
Technically we could keep in the unfolded set of also arguments of the call `(Symbol, [Expr])` but I'm worried that it could be tricked to end in an infinite loop somehow since the termination checking happens after this code is ran

I don't understand your second comment sorry

Reply:
> but I'm worried that it could be tricked to end in an infinite loop somehow

How would PLE enter an infinite loop if `unfoldedDefinitions` isn't introduced at all?

> I don't understand your second comment sorry

This ties to the motivation of `unfoldedDefinitions` which isn't clear to me yet. Symbols aren't kept in `unfoldedDefinitions` forever, they are removed with `unsetUnfolded`. And I'm asking how did you choose when to call `unsetUnfolded`.

> For the proofs I wanted to wait for some comments about the PR (both on the implementation and on the idea) hence why is marked as a draft,

I hadn't noticed this PR was marked as a draft. Collecting feedback sounds good.

Reply:
Take as an example

```haskell
{-@ LIQUID "--reflection"                @-}
{-@ LIQUID "--ple"                       @-}
{-@ LIQUID "--full"                      @-}
{-@ LIQUID "--extensionality"            @-}
{-@ LIQUID "--ple-with-undecided-guards" @-}

module Example where

{-@ reflect f @-}
f :: Int -> Int
f n = if n > 0 then f n + n else 0

{-@ reflect g @-}
g :: Int -> Int -> Int
g x y = f 3


{-@ complexProof :: { g = (\x:Int -> (\y:Int -> 3)) } @-}
complexProof :: ()
complexProof = ()

```

Even tough LH is able to infer that f is non terminating, LH will try to look for some other errors so when trying to prove `complexProof` will start with the unfolding of g so it will try to unfold f 3 so it will try to unfold then then branch of the body of f and so on till we finish the space for the call stack.

You can try it by your self by commenting away L807.

Reply:
Cool thing :) we have infinite loops only in non terminating programs

```haskell
{-@ LIQUID "--reflection"                @-}
{-@ LIQUID "--ple"                       @-}
{-@ LIQUID "--full"                      @-}
{-@ LIQUID "--extensionality"            @-}
{-@ LIQUID "--ple-with-undecided-guards" @-}

module Example where

{-@ reflect f @-}
f :: Int -> Int
f n = if n > 0 then f (n - 1) + n else 0

{-@ reflect g @-}
g :: Int -> Int -> Int
g x y = f 3


{-@ complexProof :: { g = (\x:Int -> (\y:Int -> 6)) } @-}
complexProof :: ()
complexProof = ()
```

For example this passes even without keeping track of the unfolded definitions

Reply:
Regarding set unfolded, we have to remove them from the set since we sometimes want to use the unfolding of the function in 2 different proof maybe in even different scopes

Reply:
Thanks. I had to add `{-@ lazy f @-}` to have LH accept that example.

I can see that it enters an infinite loop, and I can see why when commenting line 807. However, it passes verification if I also comment line 808 which is new, and is participating in the loop. Can the example be modified so line 808 is actually needed for verification to succeed?

Reply:
> Can the example be modified so line 808 is actually needed for verification to succeed?

Nevermind, the example in the description of the PR seems to be one such example. :thinking:

Reply:
Line 808 is what is performing the "deep" unfolding.

Lets take
```haskell
{-@ complexProof' :: { g = k } @-}
complexProof' :: ()
complexProof' = ()
```

The extensionality pass will change the proof in
```haskell
{-@ complexProof' :: { g ext$0 ext$1 = k ext$0 ext$1 } @-}
complexProof' :: ()
complexProof' = ()
```

Now what happens with the original PLE is that we have one layer of unfolding
```haskell
{-@ complexProof' :: { f ext$1 = 13 } @-}
complexProof' :: ()
complexProof' = ()
```
For`k` is enough but for `g` is not we need to unfold also `f ext$1` but the original PLE performs one unfolding ad it stops

Instead the new PLE at line 808 calls eval again and will unfold `f ext$1` to 13 making the proof automatic by LH

Reply:
> but the original PLE performs one unfolding ad it stops

I think that is a bug in PLE, it shouldn't stop. The following patch fixes an omission in the existing code, by also proving redundant the recursive call to `eval γ ctx et e3'`.

```
diff --git a/src/Language/Fixpoint/Solver/PLE.hs b/src/Language/Fixpoint/Solver/PLE.hs
index cc4644d1..1740106b 100644
--- a/src/Language/Fixpoint/Solver/PLE.hs
+++ b/src/Language/Fixpoint/Solver/PLE.hs
@@ -805,7 +805,7 @@ evalApp γ ctx e0 es et
                 ext <- isExtensionalityOn
                 when ext $ do
                   setUnfolded f
-                  _ <- eval γ ctx et e3'
+                  -- _ <- eval γ ctx et e3'
                   unsetUnfolded f
 
                 if undecidedGuards
@@ -825,7 +825,7 @@ evalApp γ ctx e0 es et
                         { evNewEqualities = S.insert (eApps e0 es, e3') (evNewEqualities st)
                         , evPendingUnfoldings = M.delete (eApps e0 es) (evPendingUnfoldings st)
                         }
-                    return (Just e2', fe)
+                    return (Just $ eApps e2' es2, fe)
          else return (Nothing, noExpand)
   where
     -- At the time of writing, any function application wrapping an
```

PLE cannot continue if the `evalApp` doesn't return the correct unfolding of the expression `eApp e0 es`. 

But this is not the end of the story for me. When tracing the above example, I found that the reflection of `g` is amiss. In the .fq file, it looks correct:
```
define Example.g (lq1 : int,  lq2 : int) : int = {((Example.f lq2))}
```
but in the `eq` variable in `evalApp` it is shows as:
```
Equ
  { eqName = "Example.g"
  , eqArgs = [("lq1##Example.g##0",FInt)]
  , eqBody = EVar "Example.f"
  , eqSort = FFunc FInt FInt, eqRec = False
  }
```
It is missing the "lq2" argument, and the eqBody should be a function application, not a variable. Has `liquid-fixpoint` eta reduced the reflection??

Reply:
`return (Just $ eApps e2' es2, fe)` makes LH perform the right amount of unfolding so that if able to verify `g = k`? How does it handles the infinite loop? I have turned out my PC so I cant test it right now but it marks `Whole.hs` as safe?

Reply:
> How does it handles the infinite loop?

PLE keeps a list of unfoldings that have been discovered (`icEquals`). I guess the infinite loop is avoided because it is crafted by always unfolding the same expression over and over. Once the unfolding enters `icEquals`, PLE doesn't continue.

I tried all the examples in this PR, and they all seem to pass.

Reply:
Can you try Whole.hs in this gist? https://gist.github.com/AlecsFerra/9d48ae08f63d8192b2267d0e378f8c09 it is rejected by the old PLE but accepted by mine

Note that we still need the code that I added for the evaluation of lambdas

Reply:
I can still make PLE diverge with
```
{-@ reflect f @-}
{-@ lazy f @-}
f :: Int -> Int
f n = if n > 0 then f (n+1) + n else 0

{-@ reflect g @-}
g :: Int -> Int -> Int
g x y = f 3

{-@ complexProof :: { g = (\x:Int -> (\y:Int -> 3)) } @-}
complexProof :: ()
complexProof = ()
```
but same as before, one still needs to use `lazy` which is an unsafe thing to do when using the function in proofs.

Reply:
If its the case then I guess we can perform the same modification in the case I added at line 868, removing the recursive call call and modifying the return

Reply:
> we can perform the same modification in the case I added at line 868

The return in that case looks correct to me.

Reply:
Yeah but I'm calling eval inside that case maybe it isn't needed if PLE decides by looking at what is returned it it wants to unfold again

Reply:
Btw regarding making PLE diverge i think its our job to make LH terminate anyway, also before I added the infinite loop check, some test in the liquidhaskell repo was entering in the same infinite loop

Reply:
Happy to discuss that further, but I don't think it affects the contribution in this PR. I'd propose to open a separate issue or PR for that sake.